tests/ui/precedence.fixed

```diff
@@ -35,3 +40,20 @@ fn main() {
     let b = 3;
     trip!(b * 8);
 }
+
+struct W(u8);
+impl Clone for W {
+    fn clone(&self) -> Self {
+        W(1)
+    }
+}
+
+fn closure_method_call() {
+    let f = move |x: W| { { x }.clone() };
+    assert!(matches!(f(W(0)), W(1)));
+    //~^^ precedence
+
+    let f = (move |x: W| -> _ { x }).clone();
+    assert!(matches!(f(W(0)), W(0)));
+    //~^^ precedence
```

Comment:
```suggestion
```

I'd prefer we not do this one.  In my model, the one that's "wrong" is the `|x| { .. }.something()` case, because it's reasonable to assume that opening brace triggers the parsing of a block, when it doesn't.  This one, by contrast, does the "right" thing, and with the annotated return type there, which then requires those braces, I think it's reasonably clear to the reader that the closure body must terminate with the closing brace.

There's also a practical reason I don't want us pushing the parens here.  I'm interested in us later implementing `IntoFuture` for thunk async closures.  This would then allow writing `async || -> _ { .. }.into_future()`, and I think it'd be better ergonomics for those to not require a set of parens around them to please clippy.

Reply:
> This one, by contrast, does the "right" thing

I find `async || -> _ { .. }.into_future()` quite ambiguous and would have a hard time parsing it if I saw it in code, so I am in favor of requiring more explicit parentheses here. Or rather, I'd probably write this as `IntoFuture::into_future(async || { ... })`.

Reply:
I don't know, I feel like `|| -> _ {}()`, which we accept without linting, feels the same to me, so I'd prefer to lint against both or neither.

And for both, I feel like they look worse in abstract form than they would in practice, e.g. with the block split over multiple lines or the return type ascribed in other ways.

Reply:
The broader point is that `|| -> _ { .. }// anything here` shouldn't feel ambiguous to us because the only thing that can follow the `-> _` is a block, so it's completely unambiguous where that block ends.

The only reason that it might feel ambiguous to us is because of what we did with `|| { ..`, where the opening brace doesn't force block parsing.  I would like us to move away from that.  Perhaps we could lint hard against anything that relies on it.  Maybe over an edition we could change the rule.

My hope is that in pushing in that direction, `|| -> _ { .. }// anything here` won't feel ambiguous to us any longer.

Reply:
> || -> _ { .. }// anything here shouldn't feel ambiguous to us because the only thing that can follow the -> _ is a block

The fact that `-> _` affects parsing like that is arcane knowledge most people don't have. I heard about it the first time ~2 weeks ago.

> The only reason that it might feel ambiguous to us (I claim) is because of what we did with || { .., where the opening brace doesn't force block parsing.  I would like us to move away from that. Perhaps we could lint hard against anything that relies on it. Maybe over an edition we could change the rule.

Ah, so you are saying we could make `|| -> _ {` actually consistent with `|| {`? Yeah that seems nice, and should definitely be possible with an edition transition.

Reply:
> normal intuitions about blocks kick in

My intuition is surprised whenever these end of block cases pop up, for example that the parens are required in

```rust
fn f() -> Y {
    (unsafe { x } as Y)
}
```

Similarly, the current parsing of `|x: String| {x}.clone()` matches my intuition